File: app/src/main/java/com/mutualmobile/praxis/ui/base/BaseActivity.java
package com.mutualmobile.praxis.ui.base;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.v7.app.AppCompatActivity;
import com.mutualmobile.praxis.BR;
import com.mutualmobile.praxis.BaseApplication;
import com.mutualmobile.praxis.injection.component.ActivityComponent;
import com.mutualmobile.praxis.injection.module.ActivityModule;
import javax.inject.Inject;

public abstract class BaseActivity<B extends ViewDataBinding, V extends MvvmView, VM extends MvvmViewModel> extends AppCompatActivity {
  protected ActivityComponent component;

  @Inject protected VM viewModel;

  protected B binding;

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    // Inject dependencies
    onComponentCreated(getComponent());
    // Bind the view and bind the viewModel to layout
    bindContentView(layoutId());
  }

  public void bindContentView(int layoutId) {
    binding = DataBindingUtil.setContentView(this, layoutId);
    viewModel.attachView((MvvmView) this);
    binding.setVariable(BR.viewModel, viewModel);
  }

  @Override protected void onDestroy() {
    super.onDestroy();
    viewModel.detachView();
  }

  public ActivityComponent getComponent() {
    if (component == null) {
      component = ((BaseApplication) getApplication()).getComponent().plusActivityComponent(new ActivityModule(this));
    }
    return component;
  }

  @LayoutRes protected abstract int layoutId();

  protected abstract void onComponentCreated(ActivityComponent component);
}

File: app/src/main/java/com/mutualmobile/praxis/ui/base/BaseActivity.kt
package com.mutualmobile.praxis.ui.base

import android.os.Bundle
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.mutualmobile.praxis.BR
import dagger.android.support.DaggerAppCompatActivity
import javax.inject.Inject

abstract class BaseActivity<B : ViewDataBinding, VM : ViewModel> : DaggerAppCompatActivity() {
  protected lateinit var binding: B
  lateinit var viewModel: VM

  @Inject
  lateinit var viewModelFactory: ViewModelProvider.Factory

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    // Bind the view and bind the viewModel to layout
    bindContentView(layoutId())
  }

  private fun bindContentView(layoutId: Int) {
    binding = DataBindingUtil.setContentView(this, layoutId)
    viewModel = ViewModelProvider(this, viewModelFactory).get(getViewModelClass())
    binding.setVariable(BR.viewModel, viewModel)
  }

  abstract fun getViewModelClass(): Class<VM>

  @LayoutRes protected abstract fun layoutId(): Int

}
